x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java

```diff
@@ -64,427 +113,1139 @@
+                    }
+                }, candidate.simplePattern.pattern(), candidate.strictGrokPattern, candidate.outputGrokPatternName);
+            }
+        }
+
+        // None of the out-of-the-box formats were close, so use the built Grok pattern and simple regex for the override
+        return new CandidateTimestampFormat(example -> Collections.singletonList(overrideFormat),
+            grokPatternAndRegex.v2(), grokPatternAndRegex.v1(), CUSTOM_TIMESTAMP_GROK_NAME);
     }
 
     /**
-     * Find the first timestamp format that matches part of the supplied value,
-     * excluding a specified number of candidate formats.
-     * @param text             The value that the returned timestamp format must exist within.
-     * @param ignoreCandidates The number of candidate formats to exclude from the search.
-     * @param requiredFormat A timestamp format that any returned match must support.
-     * @param timeoutChecker Will abort the operation if its timeout is exceeded.
+     * Find the first timestamp format that matches part or all of the supplied text.
+     * @param candidate        The timestamp candidate to consider.
+     * @param text             The text that the returned timestamp format must exist within.
+     * @param numberPosBitSet  If not <code>null</code>, each bit must be set to <code>true</code> if and only if the
+     *                         corresponding position in {@code text} is a digit.
+     * @param requireFullMatch Does the candidate have to match the entire text?
+     * @param timeoutChecker   Will abort the operation if its timeout is exceeded.
      * @return The timestamp format, or <code>null</code> if none matches.
      */
-    public static TimestampMatch findFirstMatch(String text, int ignoreCandidates, String requiredFormat, TimeoutChecker timeoutChecker) {
-        if (ignoreCandidates >= ORDERED_CANDIDATE_FORMATS.size()) {
-            return null;
-        }
-        Boolean[] quickRuleoutMatches = new Boolean[QUICK_RULE_OUT_PATTERNS.size()];
-        int index = ignoreCandidates;
-        String adjustedRequiredFormat = adjustRequiredFormat(requiredFormat);
-        for (CandidateTimestampFormat candidate : ORDERED_CANDIDATE_FORMATS.subList(ignoreCandidates, ORDERED_CANDIDATE_FORMATS.size())) {
-            if (adjustedRequiredFormat == null || candidate.jodaTimestampFormats.contains(adjustedRequiredFormat) ||
-                candidate.javaTimestampFormats.contains(adjustedRequiredFormat)) {
-                boolean quicklyRuledOut = false;
-                for (Integer quickRuleOutIndex : candidate.quickRuleOutIndices) {
-                    if (quickRuleoutMatches[quickRuleOutIndex] == null) {
-                        quickRuleoutMatches[quickRuleOutIndex] = QUICK_RULE_OUT_PATTERNS.get(quickRuleOutIndex).matcher(text).find();
+    private static TimestampMatch checkCandidate(CandidateTimestampFormat candidate, String text, BitSet numberPosBitSet,
```

Comment:
add `@Nullable` where suitable? Especially in methods with optional arguments it helps a lot when reading.